README.md
# Flutter Beacon

[![Pub](https://img.shields.io/pub/v/flutter_beacon.svg)](https://pub.dartlang.org/packages/flutter_beacon) [![GitHub](https://img.shields.io/github/license/alann-maulana/flutter_beacon.svg?color=2196F3)](https://github.com/alann-maulana/flutter_beacon/blob/master/LICENSE) [![Build](https://github.com/alann-maulana/flutter_beacon/workflows/Flutter%20CI/badge.svg)](https://github.com/alann-maulana/flutter_beacon/actions?query=workflow%3A%22Flutter+CI%22) [![Coverage Status](https://coveralls.io/repos/github/alann-maulana/flutter_beacon/badge.svg?branch=master)](https://coveralls.io/github/alann-maulana/flutter_beacon?branch=master)  [![FOSSA Status](https://app.fossa.com/api/projects/git%2Bgithub.com%2Falann-maulana%2Fflutter_beacon.svg?type=shield)](https://app.fossa.com/projects/git%2Bgithub.com%2Falann-maulana%2Fflutter_beacon?ref=badge_shield) [![codecov](https://codecov.io/gh/alann-maulana/flutter_beacon/branch/master/graph/badge.svg)](https://codecov.io/gh/alann-maulana/flutter_beacon) 

[Flutter plugin](https://pub.dartlang.org/packages/flutter_beacon/) to work with iBeacons.  

An hybrid iBeacon scanner and transmitter SDK for Flutter plugin. Supports Android API 18+ and iOS 8+.

Features:

* Automatic permission management
* Ranging iBeacons  
* Monitoring iBeacons
* Transmit as iBeacon

## Installation

Add to pubspec.yaml:

```yaml
dependencies:
  flutter_beacon: latest
```

### Setup specific for Android

For target SDK version 29+ (Android 10, 11) is necessary to add manually ```ACCESS_FINE_LOCATION```

``` 
<uses-permission android:name="android.permission.ACCESS_FINE_LOCATION"/>
```

and if you want also background scanning: 
```
<uses-permission android:name="android.permission.ACCESS_BACKGROUND_LOCATION" />
```

### Setup specific for iOS

In order to use beacons related features, apps are required to ask the location permission. It's a two step process:

1. Declare the permission the app requires in configuration files
2. Request the permission to the user when app is running (the plugin can handle this automatically)

The needed permissions in iOS is `when in use`.

For more details about what you can do with each permission, see:  
https://developer.apple.com/documentation/corelocation/choosing_the_authorization_level_for_location_services

Permission must be declared in `ios/Runner/Info.plist`:

```xml
<dict>
  <!-- When in use -->
  <key>NSLocationWhenInUseUsageDescription</key>
  <string>Reason why app needs location</string>
  <!-- Always -->
  <!-- for iOS 11 + -->
  <key>NSLocationAlwaysAndWhenInUseUsageDescription</key>
  <string>Reason why app needs location</string>
  <!-- for iOS 9/10 -->
  <key>NSLocationAlwaysUsageDescription</key>
  <string>Reason why app needs location</string>
  <!-- Bluetooth Privacy -->
  <!-- for iOS 13 + -->
  <key>NSBluetoothAlwaysUsageDescription</key>
  <string>Reason why app needs bluetooth</string>
</dict>
```

## iOS Troubleshooting

* Example code works properly only on **physical device** (bluetooth on simulator is disabled)
* How to deploy flutter app on iOS device [Instruction](https://flutter.dev/docs/get-started/install/macos)
* If example code don't works on device (beacons not appear), please make sure that you have enabled <br/> Location and Bluetooth (Settings -> Flutter Beacon) 

## How-to

Ranging APIs are designed as reactive streams.  

* The first subscription to the stream will start the ranging

### Initializing Library

```dart
try {
  // if you want to manage manual checking about the required permissions
  await flutterBeacon.initializeScanning;
  
  // or if you want to include automatic checking permission
  await flutterBeacon.initializeAndCheckScanning;
} on PlatformException catch(e) {
  // library failed to initialize, check code and message
}
```

### Ranging beacons

```dart
final regions = <Region>[];

if (Platform.isIOS) {
  // iOS platform, at least set identifier and proximityUUID for region scanning
  regions.add(Region(
      identifier: 'Apple Airlocate',
      proximityUUID: 'E2C56DB5-DFFB-48D2-B060-D0F5A71096E0'));
} else {
  // android platform, it can ranging out of beacon that filter all of Proximity UUID
  regions.add(Region(identifier: 'com.beacon'));
}

// to start ranging beacons
_streamRanging = flutterBeacon.ranging(regions).listen((RangingResult result) {
  // result contains a region and list of beacons found
  // list can be empty if no matching beacons were found in range
});

// to stop ranging beacons
_streamRanging.cancel();
```

### Monitoring beacons

```dart
final regions = <Region>[];

if (Platform.isIOS) {
  // iOS platform, at least set identifier and proximityUUID for region scanning
  regions.add(Region(
      identifier: 'Apple Airlocate',
      proximityUUID: 'E2C56DB5-DFFB-48D2-B060-D0F5A71096E0'));
} else {
  // Android platform, it can ranging out of beacon that filter all of Proximity UUID
  regions.add(Region(identifier: 'com.beacon'));
}

// to start monitoring beacons
_streamMonitoring = flutterBeacon.monitoring(regions).listen((MonitoringResult result) {
  // result contains a region, event type and event state
});

// to stop monitoring beacons
_streamMonitoring.cancel();
```

## Add foreground service
Add in `AndroidManifest.xml`: 
```xml
  <uses-permission android:name="android.permission.FOREGROUND_SERVICE" />
```
Call in `main`:

```dart
await flutterBeacon.setBackgroundService(
    true,
    const BackgroundServiceSettings(
      enabled: true,
      foregroundNotification: ForegroundNotification(
          channelId: NOTIFICATION_CHANNEL_ID,
          channelName: NOTIFICATION_CHANNEL_NAME,
          icon: NOTIFICATION_ICON,
          notificationTitle: NOTIFICATION_TITLE,
          notificationMessage: NOTIFICATION_MESSAGE,
      ),
      backgroundScanPeriod: BackgroundScanPeriod(
          backgroundBetweenScanPeriod: BETWEEN_SCAN_PERIOD,
          backgroundScanPeriod: SCAN_PERIOD
      )
    ),
  ); 
```

## Under the hood

* iOS uses native Framework [CoreLocation](https://developer.apple.com/documentation/corelocation/)
* Android uses the [Android-Beacon-Library](https://github.com/AltBeacon/android-beacon-library) ([Apache License 2.0](https://github.com/AltBeacon/android-beacon-library/blob/master/LICENSE))  

# Author

Flutter Beacon plugin is developed by Eyro Labs. You can contact me at <maulana@cubeacon.com>.


CHANGELOG.md
## UNRELEASED
* Added foreground service for android

## [0.5.1]
* Migration the example to null safety and to Android v2 embedding
* Update Android-Beacon-Library to version [2.19] with added support to Android 12 (https://github.com/AltBeacon/android-beacon-library/tree/2.19)
* Update README.md with new Android permission requirement for location (SDK 29+, Android 10, 11)

## [0.5.0]
* Migration to null safety

## [0.4.0]
* Update to Android v2 embedding
* Add broadcast Beacon feature
* Improve example app

## [0.3.0]
* Update Android-Beacon-Library to version [2.16.3](https://github.com/AltBeacon/android-beacon-library/tree/2.16.3)
* Add `BluetoothState` event channel
* Add `AuthorizationStatus` event channel [iOS only]
* Add manual check for `AuthorizationStatus`, `BluetoothState`, Location Service.
* Add opener settings for Bluetooth [Android], Location [Android] and Application [iOS].
* A lot of improvements

## [0.2.4]

* Fix bugs
* Add `close` scanning API

## [0.2.3]

* Update gradle build to latest version
* Update stop 

## [0.2.2]

* Update gradle to latest version
* Update Android-Beacon-Library to version [2.16.1](https://github.com/AltBeacon/android-beacon-library/tree/2.16.1)

## [0.2.1]

* Fix crash ranging when region not set for Android

## [0.2.0]

* Migrate to AndroidX

## [0.1.1]

* Fix region ranging for Android

## [0.1.0]

* Updating library docs
* Adding example docs

## [0.0.2]

* Adding Dart Docs

## [0.0.1]

* Initial release
* Ranging iBeacons for iOS and Android

android/src/main/java/com/flutterbeacon/FlutterBeaconPlugin.java
package com.flutterbeacon;

import android.app.Activity;
import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.RemoteException;
import android.util.Log;

import androidx.annotation.NonNull;

import org.altbeacon.beacon.BeaconManager;
import org.altbeacon.beacon.BeaconParser;
import org.altbeacon.beacon.MonitorNotifier;

import java.util.Map;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry;
import io.flutter.plugin.common.PluginRegistry.Registrar;

public class FlutterBeaconPlugin implements FlutterPlugin, ActivityAware, MethodCallHandler,
    PluginRegistry.RequestPermissionsResultListener,
    PluginRegistry.ActivityResultListener {
  
  private static final BeaconParser iBeaconLayout = new BeaconParser()
      .setBeaconLayout("m:2-3=0215,i:4-19,i:20-21,i:22-23,p:24-24");

  static final int REQUEST_CODE_LOCATION = 1234;
  static final int REQUEST_CODE_BLUETOOTH = 5678;

  private FlutterPluginBinding flutterPluginBinding;
  private ActivityPluginBinding activityPluginBinding;

  private FlutterBeaconScanner beaconScanner;
  private FlutterBeaconBroadcast beaconBroadcast;
  private FlutterPlatform platform;
  
  private BeaconManager beaconManager;
  Result flutterResult;
  private Result flutterResultBluetooth;
  private EventChannel.EventSink eventSinkLocationAuthorizationStatus;

  private MethodChannel channel;
  private EventChannel eventChannel;
  private EventChannel eventChannelMonitoring;
  private EventChannel eventChannelBluetoothState;
  private EventChannel eventChannelAuthorizationStatus;

  public FlutterBeaconPlugin() {

  }


  public static void registerWith(Registrar registrar) {
    final FlutterBeaconPlugin instance = new FlutterBeaconPlugin();
    instance.setupChannels(registrar.messenger(), registrar.activity());
    registrar.addActivityResultListener(instance);
    registrar.addRequestPermissionsResultListener(instance);
  }

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding binding) {
    this.flutterPluginBinding = binding;
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    this.flutterPluginBinding = null;
  }

  @Override
  public void onAttachedToActivity(@NonNull ActivityPluginBinding binding) {
    this.activityPluginBinding = binding;
    Log.d("BEACON","onAttachedToActivity");
    setupChannels(flutterPluginBinding.getBinaryMessenger(), binding.getActivity());
  }

  @Override
  public void onDetachedFromActivityForConfigChanges() {
    onDetachedFromActivity();
  }

  @Override
  public void onReattachedToActivityForConfigChanges(@NonNull ActivityPluginBinding binding) {
    onAttachedToActivity(binding);
  }

  @Override
  public void onDetachedFromActivity() {
    teardownChannels();
  }

  BeaconManager getBeaconManager() {
    return beaconManager;
  }

  private void setupChannels(BinaryMessenger messenger, Activity activity) {
    if (activityPluginBinding != null) {
      activityPluginBinding.addActivityResultListener(this);
      activityPluginBinding.addRequestPermissionsResultListener(this);
    }

    beaconManager = BeaconManager.getInstanceForApplication(activity.getApplicationContext());
    if (!beaconManager.getBeaconParsers().contains(iBeaconLayout)) {
      beaconManager.getBeaconParsers().clear();
      beaconManager.getBeaconParsers().add(iBeaconLayout);
    }

    platform = new FlutterPlatform(activity);
    beaconScanner = new FlutterBeaconScanner(this, activity);
    beaconBroadcast = new FlutterBeaconBroadcast(activity, iBeaconLayout);

    channel = new MethodChannel(messenger, "flutter_beacon");
    channel.setMethodCallHandler(this);

    eventChannel = new EventChannel(messenger, "flutter_beacon_event");
    eventChannel.setStreamHandler(beaconScanner.rangingStreamHandler);

    eventChannelMonitoring = new EventChannel(messenger, "flutter_beacon_event_monitoring");
    eventChannelMonitoring.setStreamHandler(beaconScanner.monitoringStreamHandler);

    eventChannelBluetoothState = new EventChannel(messenger, "flutter_bluetooth_state_changed");
    eventChannelBluetoothState.setStreamHandler(new FlutterBluetoothStateReceiver(activity));

    eventChannelAuthorizationStatus = new EventChannel(messenger, "flutter_authorization_status_changed");
    eventChannelAuthorizationStatus.setStreamHandler(locationAuthorizationStatusStreamHandler);
  }

  private void teardownChannels() {
    if (activityPluginBinding != null) {
      activityPluginBinding.removeActivityResultListener(this);
      activityPluginBinding.removeRequestPermissionsResultListener(this);
    }

    platform = null;
    beaconBroadcast = null;

    channel.setMethodCallHandler(null);
    eventChannel.setStreamHandler(null);
    eventChannelMonitoring.setStreamHandler(null);
    eventChannelBluetoothState.setStreamHandler(null);
    eventChannelAuthorizationStatus.setStreamHandler(null);

    channel = null;
    eventChannel = null;
    eventChannelMonitoring = null;
    eventChannelBluetoothState = null;
    eventChannelAuthorizationStatus = null;

    activityPluginBinding = null;
  }


  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull final Result result) {
    if (call.method.equals("enableBackground")) {
      boolean enable = call.argument("enabled");

        if (enable && !beaconManager.isAnyConsumerBound()) {
          final BackgroundSettings settings = BackgroundSettings.fromMap((Map<String,Object>) call.arguments);

          Notification.Builder builder = new Notification.Builder(flutterPluginBinding.getApplicationContext())
                  .setContentTitle(settings.foregroundNotification.notificationTitle)
                  .setContentText(settings.foregroundNotification.notificationMessage)
                  .setOnlyAlertOnce(true)
                  .setSmallIcon(flutterPluginBinding.getApplicationContext().getResources().getIdentifier(settings.foregroundNotification.icon,"drawable",flutterPluginBinding.getApplicationContext().getPackageName()));

          // create notification channel
          if(Build.VERSION.SDK_INT>=Build.VERSION_CODES.O)
          {
            NotificationChannel notificationChannel= new NotificationChannel(settings.foregroundNotification.channelId,settings.foregroundNotification.channelName, NotificationManager.IMPORTANCE_LOW);
            NotificationManager notificationManager= (NotificationManager) flutterPluginBinding.getApplicationContext().getSystemService(Context.NOTIFICATION_SERVICE);
            notificationManager.createNotificationChannel(notificationChannel);
            builder.setChannelId(notificationChannel.getId());
          }

          beaconManager.enableForegroundServiceScanning(builder.build(), 456);
          beaconManager.setEnableScheduledScanJobs(false);
          beaconManager.setBackgroundBetweenScanPeriod((long) settings.backgroundScanPeriod.backgroundBetweenScanPeriod);
          beaconManager.setBackgroundScanPeriod((long) settings.backgroundScanPeriod.backgroundScanPeriod);
        }
      result.success(true);
      return;
    }

    if (call.method.equals("initialize")) {
      if (beaconManager != null && !beaconManager.isBound(beaconScanner.beaconConsumer)) {
        Log.d("BEACON","initialize");

        this.flutterResult = result;
        this.beaconManager.bind(beaconScanner.beaconConsumer);

        return;
      }

      result.success(true);
      return;
    }

    if (call.method.equals("initializeAndCheck")) {
      initializeAndCheck(result);
      return;
    }

    if (call.method.equals("setScanPeriod")) {
      int scanPeriod = call.argument("scanPeriod");
      this.beaconManager.setForegroundScanPeriod(scanPeriod);
      try {
        this.beaconManager.updateScanPeriods();
        result.success(true);
      } catch (RemoteException e) {
        result.success(false);
      }
    }

    if (call.method.equals("setBetweenScanPeriod")) {
      int betweenScanPeriod = call.argument("betweenScanPeriod");
      this.beaconManager.setForegroundBetweenScanPeriod(betweenScanPeriod);
      try {
        this.beaconManager.updateScanPeriods();
        result.success(true);
      } catch (RemoteException e) {
        result.success(false);
      }
    }

    if (call.method.equals("setLocationAuthorizationTypeDefault")) {
      // Android does not have the concept of "requestWhenInUse" and "requestAlways" like iOS does,
      // so this method does nothing.
      // (Well, in Android API 29 and higher, there is an "ACCESS_BACKGROUND_LOCATION" option,
      //  which could perhaps be appropriate to add here as an improvement.)
      result.success(true);
      return;
    }

    if (call.method.equals("authorizationStatus")) {
      result.success(platform.checkLocationServicesPermission() ? "ALLOWED" : "NOT_DETERMINED");
      return;
    }

    if (call.method.equals("checkLocationServicesIfEnabled")) {
      result.success(platform.checkLocationServicesIfEnabled());
      return;
    }

    if (call.method.equals("bluetoothState")) {
      try {
        boolean flag = platform.checkBluetoothIfEnabled();
        result.success(flag ? "STATE_ON" : "STATE_OFF");
        return;
      } catch (RuntimeException ignored) {

      }

      result.success("STATE_UNSUPPORTED");
      return;
    }

    if (call.method.equals("requestAuthorization")) {
      if (!platform.checkLocationServicesPermission()) {
        this.flutterResult = result;
        platform.requestAuthorization();
        return;
      }

      // Here, location services permission is granted.
      //
      // It's possible location permission was granted without going through
      // our onRequestPermissionsResult() - for example if a different flutter plugin
      // also requested location permissions, we could end up here with
      // checkLocationServicesPermission() returning true before we ever called requestAuthorization().
      //
      // In that case, we'll never get a notification posted to eventSinkLocationAuthorizationStatus
      //
      // So we could could have flutter code calling requestAuthorization here and expecting to see
      // a change in eventSinkLocationAuthorizationStatus but never receiving it.
      //
      // Ensure an ALLOWED status (possibly duplicate) is posted back.
      if (eventSinkLocationAuthorizationStatus != null) {
        eventSinkLocationAuthorizationStatus.success("ALLOWED");
      }

      result.success(true);
      return;
    }

    if (call.method.equals("openBluetoothSettings")) {
      if (!platform.checkBluetoothIfEnabled()) {
        this.flutterResultBluetooth = result;
        platform.openBluetoothSettings();
        return;
      }

      result.success(true);
      return;
    }

    if (call.method.equals("openLocationSettings")) {
      platform.openLocationSettings();
      result.success(true);
      return;
    }

    if (call.method.equals("openApplicationSettings")) {
      result.notImplemented();
      return;
    }

    if (call.method.equals("close")) {
      Log.d("BEACON","close");
      if (beaconManager != null) {
        beaconScanner.stopRanging();
        beaconManager.removeAllRangeNotifiers();
        beaconScanner.stopMonitoring();
        beaconManager.removeAllMonitorNotifiers();
        if (beaconManager.isBound(beaconScanner.beaconConsumer)) {
          beaconManager.unbind(beaconScanner.beaconConsumer);
        }
      }
      result.success(true);
      return;
    }

    if (call.method.equals("startBroadcast")) {
      beaconBroadcast.startBroadcast(call.arguments, result);
      return;
    }

    if (call.method.equals("stopBroadcast")) {
      beaconBroadcast.stopBroadcast(result);
      return;
    }

    if (call.method.equals("isBroadcasting")) {
      beaconBroadcast.isBroadcasting(result);
      return;
    }

    if (call.method.equals("isBroadcastSupported")) {
      result.success(platform.isBroadcastSupported());
      return;
    }

    result.notImplemented();
  }

  private void initializeAndCheck(Result result) {
    if (platform.checkLocationServicesPermission()
        && platform.checkBluetoothIfEnabled()
        && platform.checkLocationServicesIfEnabled()) {
      if (result != null) {
        result.success(true);
        return;
      }
    }

    flutterResult = result;
    if (!platform.checkBluetoothIfEnabled()) {
      platform.openBluetoothSettings();
      return;
    }

    if (!platform.checkLocationServicesPermission()) {
      platform.requestAuthorization();
      return;
    }

    if (!platform.checkLocationServicesIfEnabled()) {
      platform.openLocationSettings();
      return;
    }

    if (beaconManager != null && !beaconManager.isBound(beaconScanner.beaconConsumer)) {
      if (result != null) {
        this.flutterResult = result;
      }

      beaconManager.bind(beaconScanner.beaconConsumer);
      return;
    }

    if (result != null) {
      result.success(true);
    }
  }

  private final EventChannel.StreamHandler locationAuthorizationStatusStreamHandler = new EventChannel.StreamHandler() {
    @Override
    public void onListen(Object arguments, EventChannel.EventSink events) {
      eventSinkLocationAuthorizationStatus = events;
    }

    @Override
    public void onCancel(Object arguments) {
      eventSinkLocationAuthorizationStatus = null;
    }
  };

  // region ACTIVITY CALLBACK
  @Override
  public boolean onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
    if (requestCode != REQUEST_CODE_LOCATION) {
      return false;
    }

    boolean locationServiceAllowed = false;
    if (permissions.length > 0 && grantResults.length > 0) {
      String permission = permissions[0];
      if (!platform.shouldShowRequestPermissionRationale(permission)) {
        int grantResult = grantResults[0];
        if (grantResult == PackageManager.PERMISSION_GRANTED) {
          //allowed
          locationServiceAllowed = true;
        }
        if (eventSinkLocationAuthorizationStatus != null) {
          // shouldShowRequestPermissionRationale = false, so if access wasn't granted, the user clicked DENY and checked DON'T SHOW AGAIN
          eventSinkLocationAuthorizationStatus.success(locationServiceAllowed ? "ALLOWED" : "DENIED");
        }
      }
      else {
        // shouldShowRequestPermissionRationale = true, so the user has clicked DENY but not DON'T SHOW AGAIN, we can possibly prompt again
        if (eventSinkLocationAuthorizationStatus != null) {
          eventSinkLocationAuthorizationStatus.success("NOT_DETERMINED");
        }
      }
    }
    else {
      // Permission request was cancelled (another requestPermission active, other interruptions), we can possibly prompt again
      if (eventSinkLocationAuthorizationStatus != null) {
        eventSinkLocationAuthorizationStatus.success("NOT_DETERMINED");
      }
    }

    if (flutterResult != null) {
      if (locationServiceAllowed) {
        flutterResult.success(true);
      } else {
        flutterResult.error("Beacon", "location services not allowed", null);
      }
      this.flutterResult = null;
    }

    return locationServiceAllowed;
  }

  @Override
  public boolean onActivityResult(int requestCode, int resultCode, Intent intent) {
    boolean bluetoothEnabled = requestCode == REQUEST_CODE_BLUETOOTH && resultCode == Activity.RESULT_OK;

    if (bluetoothEnabled) {
      if (!platform.checkLocationServicesPermission()) {
        platform.requestAuthorization();
      } else {
        if (flutterResultBluetooth != null) {
          flutterResultBluetooth.success(true);
          flutterResultBluetooth = null;
        } else if (flutterResult != null) {
          flutterResult.success(true);
          flutterResult = null;
        }
      }
    } else {
      if (flutterResultBluetooth != null) {
        flutterResultBluetooth.error("Beacon", "bluetooth disabled", null);
        flutterResultBluetooth = null;
      } else if (flutterResult != null) {
        flutterResult.error("Beacon", "bluetooth disabled", null);
        flutterResult = null;
      }
    }

    return bluetoothEnabled;
  }
  // endregion
}
